### DISCUSSION TITLE
Am I able to use the authorisation and authentication features with Next.js?

### DISCUSSION THREAD

**Discussion by Adrian-Samuel** [keystonejs/keystone#7763] [Questions]
I have three questions, likely interrelated:

1. KeystoneJS provides session management, along with some RBAC-like functionality. Is this limited to controlling what can be seen on the CMS backend or does this extend as a middleware for the data that the APIs display (for the Front-End)

2. Do I still need NextAuth to manage auth on my Next.js front-end or does Keystone's auth eliminate the need for that?

3. If I want a super admin to be able to login to the CMS backend and amend some things, how can I allow them to also access a unique admin panel on the Front-End without having to issue two separate sets of credentials?

I reckon these questions are interlinked and things are just jumbled up in my mind as I've never used a CMS before, just front-ends with custom backends.

**Comment by flexdinesh** [comment]
Hey @Adrian-Samuel! 

1. How keystone works:
With Keystone you write schema for your data models and you get a GraphQL API ready made for you. The admin ui uses this same graphql API to show you the data. So any access control you set (Eg. only admins can access a particular list) will be applicable to the queries/mutations in the GraphQL API and that's how they are access prevented in the admin UI. 

2. Next.js frontend and session management
If you're building a Next.js frontend you will consume the GraphQL API and build a signin page and use the authentication query to authenticate. So your session is managed exactly the same way as it is in admin UI. Easiest implementation is to set the authentication token in the header/cookie in all your graphql requests.

3. Super admin
You can setup role based access to your lists and super admins will be able to login to the CMS and Next.js frontend using the same credentials and the users who don't have the role will not have access to the data either in the CMS or in your frontend (via graphql queries).

**Reply by Adrian-Samuel** [reply]
@flexdinesh thank you so much for your response! This has been super helpful!

I have one more question, how would I be able to setup dynamic access control?
I want for example to grant/remove access for certain objects based on a permissioning system that depends on the behaviour of users in my UI. How would I accomplish that?

**Reply by flexdinesh** [reply]
You can make access control dynamic by making your role a separate list instead of a static enum — Example here https://github.com/keystonejs/prisma-day-2021-workshop/blob/fd191293619f93a4c70b8f4fcc98348dfc61ce1d/schema/users.ts#L74

Create a bunch of roles and and you could implement some sort of logic to promote/demote user roles based on their behaviour.
